modules/c++/mem/source/ScratchMemory.cpp

```diff
@@ -75,9 +75,9 @@ void ScratchMemory::release(const std::string& key)
                                                                mKeyOrder.end(),
                                                                key);
         std::vector<std::string>::iterator nextKeyIter = mKeyOrder.erase(keyIter);
+        const std::string nextKey = *nextKeyIter;
         mKeyOrder.push_back(key);
 
-        const std::string nextKey = *nextKeyIter;
```

Comment:
There are references to `nextKeyIter` later in this function, which may be invalidated when `mKeyOrder.push_back(key)` reallocates. I suggest replacing all of the `*nextKeyIter` after the `push_back` with `nextKey`.

Reply:
Good catch! There was a spot below where it was actually relying on the iterator in a messy way, so I put the `push_back` operation at the start. Then when it searches the vector it should still find the original element.

I'm not sure how that behaves if the key was not in the vector originally. I didn't see any checks for if it wasn't found. Should that be able to happen?

Reply:
It looks to me like if the key is in `mSegments`, it must also be in `mKeyOrder`.

I think the first line in the `release` method `mSegments.find(key)` should probably either return or throw if the key is not found.

Reply:
How's this?

Reply:
That looks good to me.